daemon/algod/api/server/v2/handlers.go

```diff
@@ -1753,62 +1754,93 @@ func (v2 *Handlers) GetApplicationByID(ctx echo.Context, applicationID uint64) e
-		}
-		return algodMax + 1 // API limit dominates.  Increments by 1 to test if more than max supported results exist.
+func applicationBoxesMaxKeys(requestedMax uint64, algodMax uint64) int {
+	if requestedMax == 0 || requestedMax > math.MaxInt {
+		requestedMax = math.MaxInt
 	}
-
-	if requestedMax <= algodMax || algodMax == 0 {
-		return requestedMax // requested limit dominates
+	if algodMax == 0 || algodMax > math.MaxInt {
+		algodMax = math.MaxInt
 	}
-
-	return algodMax + 1 // API limit dominates.  Increments by 1 to test if more than max supported results exist.
+	return int(min(requestedMax, algodMax))
 }
 
-// GetApplicationBoxes returns the box names of an application
+// GetApplicationBoxes returns the boxes of an application
 // (GET /v2/applications/{application-id}/boxes)
 func (v2 *Handlers) GetApplicationBoxes(ctx echo.Context, applicationID uint64, params model.GetApplicationBoxesParams) error {
 	appIdx := basics.AppIndex(applicationID)
 	ledger := v2.Node.LedgerForAPI()
-	lastRound := ledger.Latest()
-	keyPrefix := apps.MakeBoxKey(uint64(appIdx), "")
 
 	requestedMax, algodMax := nilToZero(params.Max), v2.Node.Config().MaxAPIBoxPerApplication
 	max := applicationBoxesMaxKeys(requestedMax, algodMax)
 
-	if max != math.MaxUint64 {
-		record, _, _, err := ledger.LookupAccount(ledger.Latest(), appIdx.Address())
+	values := nilToZero(params.Values)
+
+	// We'll need to convert between "KV" names and "Box" names, so keep track
+	// of how much gets tacked on to make the kv name.
+	kvPrefix := apps.MakeBoxKey(uint64(appIdx), "")
+	kvPrefixLen := len(kvPrefix)
+
+	prefix := nilToZero(params.Prefix)
+	if len(prefix) > 0 {
+		cb, err := apps.NewAppCallBytes(prefix)
 		if err != nil {
-			return internalError(ctx, err, errFailedLookingUpLedger, v2.Log)
+			return badRequest(ctx, err, err.Error(), v2.Log)
+		}
+		rawPrefix, err := cb.Raw()
+		if err != nil {
+			return badRequest(ctx, err, err.Error(), v2.Log)
+		}
+		prefix = kvPrefix + string(rawPrefix)
+	}
+
+	next := nilToZero(params.Next)
+	if len(next) > 0 {
+		cb, err := apps.NewAppCallBytes(next)
+		if err != nil {
+			return badRequest(ctx, err, err.Error(), v2.Log)
 		}
-		if record.TotalBoxes > max {
+		rawNext, err := cb.Raw()
+		if err != nil {
+			return badRequest(ctx, err, err.Error(), v2.Log)
+		}
+		next = kvPrefix + string(rawNext)
+	}
+
+	round, boxes, nextToken, err := ledger.LookupKeysByPrefix(prefix, next, max, 1_000_000, values)
+	if err != nil {
+		return internalError(ctx, err, errFailedLookingUpLedger, v2.Log)
+	}
+
+	if nextToken != "" {
+		// Preserve existing failure behavior if caller is not using `next`.
+		if params.Next == nil {
 			return ctx.JSON(http.StatusBadRequest, model.ErrorResponse{
 				Message: "Result limit exceeded",
-				Data: &map[string]interface{}{
+				Data: &map[string]any{
 					"max-api-box-per-application": algodMax,
 					"max":                         requestedMax,
-					"total-boxes":                 record.TotalBoxes,
```

Comment:
why drop this part of the result?

Reply:
Because it's sort of meaningless now. We used to return it because the call was always for every box in the app. But now there are various ways of selecting only what you want (prefix, next).  This was a weird compromise we reached when we had an unpaginated hard limit - we wanted to let people know why their call failed.

I'd be shocked if anyone dug into those error values, I'm not sure the sdks even support getting them easily.

Reply:
Oh, also we don't even lookup `record` any more.